File: Data Structures & Algorithms/top-k-elements-in-list/submission-2.cpp
class Solution {
public:
    vector<int> topKFrequent(vector<int>& nums, int k) {
        unordered_map<int, int> frequncy_map; // number, fequency
        for (int n : nums){
            frequncy_map[n] = 1 + frequncy_map[n];
        }

        vector<vector<int>> frequncy_array(nums.size() + 1); // frequency, numbers

        for (const auto& item : frequncy_map){
            frequncy_array[item.second].push_back(item.first);
        }

        vector<int> res;
        for (int i = frequncy_array.size() - 1; i < 0; --i) {
            for (int n : frequncy_array[i]) {
                res.push_back(n);
                if (res.size() == k) return res;
            }
        }
    }
};

File: Data Structures & Algorithms/three-integer-sum/submission-2.py
class Solution:
    def threeSum(self, nums: List[int]) -> List[List[int]]:
        out = []
        nums.sort() # first sort numbers
        
        for i, a in enumerate(nums):
            if i > 0 and a == nums[i - 1]:
                continue
            
            start, end = i + 1, len(nums) - 1

            target_sum = 0 - a
            
            while start < end:
                if nums[start] + nums[end] > target_sum:
                    end -= 1
                elif nums[start] + nums[end] < target_sum:
                    start += 1
                else: 
                    out.append([a, nums[start], nums[end]])
                    start += 1
                    while nums[start] == nums[start - 1] and start < end:
                        start += 1

        return out
        